staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go

```diff
@@ -166,3 +166,51 @@ type TracingConfiguration struct {
 	// Defaults to 0.
 	SamplingRatePerMillion *int32
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityAndFairnessConfiguration holds API Priority and Fairness parameters.
+type PriorityAndFairnessConfiguration struct {
+	metav1.TypeMeta
+
+	// WorkEstimatorConfig holds work estimator parameters.
+	WorkEstimatorConfiguration WorkEstimatorConfiguration
+}
+
+// WorkEstimatorConfiguration holds work estimator parameters.
+type WorkEstimatorConfiguration struct {
+	// ListWorkEstimatorConfiguration holds work estimator parameters related to list requests.
+	ListWorkEstimatorConfiguration ListWorkEstimatorConfiguration
+	// MutatingWorkEstimatorConfiguration holds work estimator
+	// parameters related to watches of mutating objects.
+	MutatingWorkEstimatorConfiguration MutatingWorkEstimatorConfiguration
+
+	// MaximumSeats is the maximum number of seats a request can occupy
+	//
+	// NOTE: work_estimate_seats_samples metric uses the value of maximumSeats
+	// as the upper bound, so when we change maximumSeats we should also
+	// update the buckets of the metric.
+	MaximumSeats uint64
+}
+
+// ListWorkEstimatorConfiguration holds work estimator parameters related to list requests.
+type ListWorkEstimatorConfiguration struct {
+	// ObjectsPerSeat is a number of objects from a single list request per seat.
+	ObjectsPerSeat float64
+}
+
+// MutatingWorkEstimatorConfiguration holds work estimator
+// parameters related to watches of mutating objects.
+type MutatingWorkEstimatorConfiguration struct {
+	// Enabled enables mutating work estimator.
+	// This option will be removed once we have better proof from
+	// production usecases that the feature works as expected.
+	Enabled bool
+	// EventAdditionalDuration is the additional amount of time that a mutating request occupies seats
+	// to account for the work involved in sending the corresponding WATCH notifications.
+	EventAdditionalDuration metav1.Duration
+	// WatchesPerSeat is the ratio between the number of WATCHes that will get notified and
+	// the number of seats occupied for EventAdditionalDuration to account for the work of
+	// sending those notifications.
+	WatchesPerSeat float64
```

Comment:
would the values chosen for these noticeably impact the behavior of the same P&F config objects? if so, what guidance would we give for P&F API authors to write portable config they could rely on working well across clusters without visibility to the server-side config?

Reply:
As discussed offline, currently we also have `max-requsts-inflights` that can affect how APF is behaving and this is expected as different machine types can handle different amount of load. That's also one of reasons I prefer having this config as parameter rather than api object - this might be useful for example when you have different machine types in control plane or if etcd is not on the same machine as apiserver etc. I like to think about APF configuration this way: You configure parameters like `max-requests-inflights` or `EventAdditionalDuration/WatchesPerSeat` etc to ensure that apiserver can handle specific amount of load depending on vm machine type/networking setup etc and then once you ensure reliability you configure P&F API to ensure fairness, which corresponds to  goals of P&F from KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1040-priority-and-fairness/README.md#goals
```
This KEP introduces a new functionality that is supposed to address the problems above. The goals/requirements for the solution (in the priority order) are the following:
- overload protection
- fairness
- throughput
```